vlasovsolver/gpu_pitch_angle_diffusion.cpp

```diff
@@ -0,0 +1,799 @@
+   const int mui = std::max(0,std::min(muindex,nbins_mu-1));
+   const int vi = std::max(0,std::min(Vindex,nbins_v-1));
+   dfdt = GPUCELLMUSPACE(dev_dfdt_mu,cellIdx,vi,mui); // dfdt_mu was scaled back down by 2pi*v^2 on creation
+
+   // Update cell value, ensuring result is non-negative
+   Realf NewCellValue    = (dev_velocityBlockContainer[cellIdx]->getData()[velocityIdx*WID3+k*WID2+j*WID+i]) + dfdt * dev_Ddt[remappedCellIdx];
+   const bool lessZero = (NewCellValue < 0.0);
+   NewCellValue = lessZero ? 0.0 : NewCellValue;
+   dev_velocityBlockContainer[cellIdx]->getData()[velocityIdx*WID3+k*WID2+j*WID+i] = NewCellValue;
+}
+
+__global__ void computeDerivativesCFLDdt_kernel(
+   size_t *dev_smallCellIdxArray,
+   int *dev_fcount,
+   Realf *dev_fmu,
+   Real *dev_nu0Values,
+   Realf *dev_dfdt_mu,
+   int *dev_cellIdxKeys,
+   Real *dev_potentialDdtValues,
+   Realf *dev_sparsity,
+   Real dVbins,
+   const Real dmubins,
+   const Real epsilon,
+   Realf PADCFL,
+   int nbins_v,
+   int nbins_mu,
+   int blocksPerVelocityCell,
+   int lastBlockSize
+   ){
+
+   int spatialBlockIndex = blockIdx.x/blocksPerVelocityCell; // Corresponds to index spatial and velocity blocks
+   int indexInsideBlock = blockIdx.x%blocksPerVelocityCell;
+   int nextIndexInsideBlock = (blockIdx.x+1)%blocksPerVelocityCell;
+   int idx = indexInsideBlock*blockDim.x+threadIdx.x;
+   int threadIndex = threadIdx.x;
+   int indv = idx%nbins_v;
+   int indmu = (idx/nbins_v)%nbins_mu;
+   size_t cellIdx = dev_smallCellIdxArray[spatialBlockIndex];
+   
+   // Initiate shared memory values
+   extern __shared__ Real localDdtValues[];
+   localDdtValues[threadIndex] = std::numeric_limits<Real>::max();
+
+   __syncthreads();
+
+   if(nextIndexInsideBlock != 0 || threadIndex < lastBlockSize){
+
+      // Search limits for how many cells in mu-direction should be max evaluated when searching for a near neighbour?
+      // Assuming some oversampling; changing these values may result in method breaking at very small plasma frame velocities.
+      const int rlimit = nbins_mu-1;
+      const int llimit = 0;
+
+      int cLeft = 0;
+      int cRight = 0;
+
+      // !!! DANGER, WARP DIVERGENCE AHEAD !!!
+      // TODO: Can this be avoided?
+
+      if (indmu != nbins_mu-1) {
+         for(cRight = 1; indmu + cRight < rlimit; cRight++){
+            if(GPUCELLMUSPACE(dev_fcount,cellIdx,indv,indmu + cRight) != 0){
+               break;
+            }
+         }
+         if( (GPUCELLMUSPACE(dev_fcount,cellIdx,indv,indmu + cRight) == 0) && (indmu + cRight == rlimit) ) {
+            cRight  = 0;
+         }
+      }
+      if (indmu != 0) {
+         for(cLeft = 1; indmu - cLeft > llimit; cLeft++){
+            if(GPUCELLMUSPACE(dev_fcount,cellIdx,indv,indmu - cLeft) != 0){
+               break;
+            }
+         }
+         if( (GPUCELLMUSPACE(dev_fcount,cellIdx,indv,indmu - cLeft) == 0) && (indmu - cLeft == llimit) ) {
+            cLeft  = 0;
+         }
+      }
+
```

Comment:
if you want, you can place a sync here so warp divergence is contained to as short a region as possible.